src/shared/blockdev-util.c

```diff
@@ -476,3 +479,59 @@ int block_device_remove_partition(int fd, const char *name, int nr) {
 
         return RET_NERRNO(ioctl(fd, BLKPG, &ba));
 }
+
+int block_device_remove_all_partitions(int fd) {
+        int r;
+        struct stat stat;
+        _cleanup_(sd_device_unrefp) sd_device *dev = NULL;
+        _cleanup_(sd_device_enumerator_unrefp) sd_device_enumerator *e = NULL;
+        sd_device *part;
+
+        if (fstat(fd, &stat) < 0)
+                return -errno;
+
+        r = sd_device_new_from_devnum(&dev, 'b', stat.st_rdev);
+        if (r < 0)
+                return r;
+
+        r = sd_device_enumerator_new(&e);
+        if (r < 0)
+                return r;
+
+        r = sd_device_enumerator_add_match_parent(e, dev);
+        if (r < 0)
+                return r;
+
+        r = sd_device_enumerator_add_match_subsystem(e, "block", true);
+        if (r < 0)
+                return r;
+
+        r = sd_device_enumerator_add_match_property(e, "DEVTYPE", "partition");
+        if (r < 0)
+                return r;
+
+        FOREACH_DEVICE(e, part) {
+                const char *v, *devname;
+                int nr;
+
+                r = sd_device_get_devname(part, &devname);
+                if (r < 0)
+                        return r;
+
+                r = sd_device_get_property_value(part, "PARTN", &v);
+                if (r < 0)
+                        return r;
+
+                r = safe_atoi(v, &nr);
+                if (r < 0)
+                        return r;
+
+                r = block_device_remove_partition(fd, devname, nr);
```

Comment:
Hmm, shouldn't we handle gracefully if this returns ENODEV? (or whatever it returns when a partition is gone already?) That way it should be safe to call this when racing against the same code in the kernel?

Reply:
Should we do something with EBUSY as well here? `loop_device_unref()` already has some nasty EBUSY handling, not sure if we need that here as well

Reply:
makes sense to me

Reply:
Updated to not fail on the first error, but keep track of it and still try to remove the remaining partitions before returning the error.